File: Quasics2021Code/MaeTester/src/main/cpp/subsystems/Intake.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/Intake.h"

#include "Constants.h"
#include "LoggingUtils.h"

///////////////////////////////////////////////////////////////////////////////
// Conditional compilation flags start here.

// DEFINE this if we want to log the state reported by the ball sensor (limit
// switch or beam break sensor).
#undef LOG_BALL_SENSOR_STATE

// Conditional compilation flags end here.
///////////////////////////////////////////////////////////////////////////////

Intake::Intake()
    : intakeMotor(CANBusIds::VictorSPXIds::IntakeMotor),
      conveyorMotor(CANBusIds::VictorSPXIds::ConveyorMotor) {
#if defined(INTAKE_USES_LIMIT_SWITCH)
  conveyorLimitSwitch.reset(
      new frc::DigitalInput(DigitalIOMappings::IntakeLimitSwitch));
#elif defined(INTAKE_USES_BEAM_SENSOR)
  conveyorBeamSensor.reset(
      new frc::DigitalInput(DigitalIOMappings::ConveyorBeamSensor));
#endif
  SetSubsystem("Intake");
}

// This method will be called once per scheduler run
void Intake::Periodic() {
#if defined(LOG_BALL_SENSOR_STATE)
  static double oldValue = -1;
  double sensorValue = -1;

#if defined(INTAKE_USES_LIMIT_SWITCH)
  static const std::string sensorName("Limit switch");
  sensorValue = conveyorLimitSwitch->Get();
#elif defined(INTAKE_USES_BEAM_SENSOR)
  static const std::string sensorName("Beam sensor");
  sensorValue = conveyorBeamSensor->Get();
#endif

  if (sensorValue != oldValue) {
    // Report status only when it actually changes (to keep us from flooding the
    // console with output).
    std::cout << sensorName << " is " << (sensorValue ? "" : "not ") << "open ("
              << sensorValue << ")" << std::endl;
    oldValue = sensorValue;
  }
#endif
}

// Intakes the ball at 1/4 speed. Theoretically, when limit switch is hit, both
// Intake and Conveyor will stop.
void Intake::IntakeBallOn() {
  if (!IsBallInChamber()) {
    conveyorMotor.Set(MOTOR_FAST_POWER);
    intakeMotor.Set(MOTOR_SLOW_POWER);
  } else {
    intakeMotor.Set(MOTOR_OFF_POWER);
    conveyorMotor.Set(MOTOR_OFF_POWER);
  }
}
// Automatic Intake: if limit switch is hit, only conveyor will disable.
// Otherwise, will always run.
// TODO: Fix this: it shouldn't be written like this!!!!!
// TODO(matt): Explain to people why.
void Intake::IntakeCellsAuto() {
  intakeMotor.Set(MOTOR_FAST_POWER);
  if (IsBallInChamber()) {
    conveyorMotor.Set(MOTOR_SLOW_POWER);
  } else {
    conveyorMotor.Set(MOTOR_OFF_POWER);
  }
}

// Stops the intake of ball.
void Intake::IntakeBallOff() {
  intakeMotor.Set(MOTOR_OFF_POWER);
  conveyorMotor.Set(MOTOR_OFF_POWER);
}

void Intake::ConveyBallOn() {
  conveyorMotor.Set(MOTOR_FULL_POWER);
}

void Intake::ConveyBallReverse() {
  conveyorMotor.Set(-MOTOR_SLOW_POWER);
}

void Intake::ConveyBallOff() {
  conveyorMotor.Set(MOTOR_OFF_POWER);
}

void Intake::OnlyIntakeOn() {
  intakeMotor.Set(MOTOR_FAST_POWER);
}

void Intake::OnlyIntakeReverse() {
  intakeMotor.Set(-MOTOR_FAST_POWER);
}

void Intake::OnlyIntakeOff() {
  intakeMotor.Set(MOTOR_OFF_POWER);
}

bool Intake::IsBallInChamber() {
#if defined(INTAKE_USES_LIMIT_SWITCH)
  return !conveyorLimitSwitch->Get();
#elif defined INTAKE_USES_BEAM_SENSOR
  return !conveyorBeamSensor->Get();
#else
  LOG_EVERY_N_TIMES(50,
                    "Checking for ball in chamber, but no sensor is enabled!");
  return false;
#endif
}

void Intake::SetBallPickupSpeed(double percent) {
  // Cap the value of percent to -1.0 to +1.0.
  double useSpeed = std::max(-1.0, std::min(1.0, percent));
  intakeMotor.Set(useSpeed);
}

void Intake::SetConveyorSpeed(double percent) {
  // Cap the value of percent to -1.0 to +1.0.
  double useSpeed = std::max(-1.0, std::min(1.0, percent));
  conveyorMotor.Set(useSpeed);
}

File: Quasics2024Code/MaeTester2024-Java/src/main/java/frc/robot/subsystems/Intake.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants;

public class Intake extends SubsystemBase {
  public static final double MOTOR_OFF_POWER = 0;
  public static final double MOTOR_SLOW_POWER = 0.25;
  public static final double MOTOR_FAST_POWER = 0.75;
  public static final double MOTOR_FULL_POWER = 1.0;

  private WPI_VictorSPX m_intakeMotor =
      new WPI_VictorSPX(Constants.CANBusIds.VictorSPXIds.IntakeMotor);
  private WPI_VictorSPX m_conveyorMotor =
      new WPI_VictorSPX(Constants.CANBusIds.VictorSPXIds.ConveyorMotor);

  /** Creates a new Intake. */
  public Intake() { setSubsystem("Intake"); }

  @Override
  public void periodic() {
    // This method will be called once per scheduler run
  }

  // Intakes the ball at 1/4 speed. Theoretically, when limit switch is hit,
  // both Intake and Conveyor will stop.
  // TODO: Fix this: it shouldn't be written like this!!!!!
  // TODO(matt): Explain to people why.
  public void intakeBallOn() {
    if (!isBallInChamber()) {
      m_conveyorMotor.set(MOTOR_FAST_POWER);
      m_intakeMotor.set(MOTOR_SLOW_POWER);
    } else {
      m_intakeMotor.set(MOTOR_OFF_POWER);
      m_conveyorMotor.set(MOTOR_OFF_POWER);
    }
  }

  public void intakeBallOff() {
    m_intakeMotor.set(MOTOR_OFF_POWER);
    m_conveyorMotor.set(MOTOR_OFF_POWER);
  }

  public void conveyBallOn() { m_conveyorMotor.set(MOTOR_FULL_POWER); }
  public void conveyBallOnSlow() { m_conveyorMotor.set(MOTOR_SLOW_POWER); }
  public void conveyBallReverse() { m_conveyorMotor.set(-MOTOR_SLOW_POWER); }
  public void conveyBallOff() { m_conveyorMotor.set(MOTOR_OFF_POWER); }

  public void onlyIntakeOn() { m_intakeMotor.set(MOTOR_FAST_POWER); }
  public void onlyIntakeReverse() { m_intakeMotor.set(-MOTOR_FAST_POWER); }
  public void onlyIntakeOff() { m_intakeMotor.set(MOTOR_OFF_POWER); }

  /** TODO: Implement this using beam break sensor. */
  public boolean isBallInChamber() { return false; }

  /**
   * Sets the speed for the ball pick-up mechanism.
   * @param percent
   *           a value from 1.0 (full forward) to -1.0 (full reverse); 0 ==
   *           stop.
   */
  public void setBallPickupSpeed(
      double percent) { // Cap the value of percent to -1.0 to +1.0.
    double useSpeed = Math.max(-1.0, Math.min(1.0, percent));
    m_intakeMotor.set(useSpeed);
  }

  /**
   * Sets the speed for the conveyor used to transfer balls from the
   * "hopper" to the shooter.
   *
   * @param percent
   *           a value from 1.0 (full forward) to -1.0 (full reverse); 0 ==
   *           stop.
   */
  void setConveyorSpeed(
      double percent) { // Cap the value of percent to -1.0 to +1.0.
    double useSpeed = Math.max(-1.0, Math.min(1.0, percent));
    m_conveyorMotor.set(useSpeed);
  }
}
